lib/compiler/src/beam_trim.erl

```diff
@@ -151,15 +151,18 @@ construct_recipes(_, _, _, Acc) ->
     Acc.
 
 take_last_dead(L) ->
-    take_last_dead_1(reverse(L)).
-
-take_last_dead_1([{live,_}|Is]) ->
-    take_last_dead_1(Is);
-take_last_dead_1([{kill,Reg}|Is]) ->
-    {Reg,reverse(Is)};
-take_last_dead_1([{dead,Reg}|Is]) ->
-    {Reg,reverse(Is)};
-take_last_dead_1(_) -> none.
+    take_last_dead_1(L, []).
+
+take_last_dead_1([{live,_}|T], Acc) ->
+    take_last_dead_1(T, Acc);
+take_last_dead_1([{kill,Reg}|T], Acc) ->
+    {Reg,reverse(T, Acc)};
+take_last_dead_1([{dead,Reg}|T], Acc) ->
+    {Reg,reverse(T, Acc)};
+take_last_dead_1([], _) ->
+    none;
+take_last_dead_1([H|T], Acc) ->
+    take_last_dead_1(T, [H|Acc]).
```

Comment:
The semantics of this function has changed; you're not taking the last `kill`/`dead` anymore, but the _first_, resulting in a much weaker optimization.

Try running `scripts/diffable 0` just before the changes in this branch, and then `scripts/diffable 1` afterwards (don't forget to `make compiler` in between). This will create two folders with assembly listings for all of OTP before and after the change. Ideally, when optimizing a compiler pass a comparison of these (`diff -u 0 1 | grep ^diff`) should only show a change in the assembly listing of the pass you've optimized.